### ISSUE TITLE
S4035 reports false positive if Equals<T> is protected (e.g., in Rider generated code)

### ISSUE THREAD

**Issue by mpcleancode** [SonarSource/sonar-dotnet#4000]
### Description

The Rule S4035 reports a false positive for the code below. This is the equality members pattern generated by Rider. The pattern is safe, since the Equals(T obj) method is protected. Thus, it can not be called from another class (except a child class).

IEquatable<T> is not implemented, but S4035 reports a false positive here.

### Repro steps

```
public class TestClass
{
    public string Test { get; set; } = "Test";

    public override bool Equals(object? obj)
    {
        if (ReferenceEquals(null, obj))
        {
            return false;
        }

        if (ReferenceEquals(this, obj))
        {
            return true;
        }

        if (obj.GetType() != this.GetType())
        {
            return false;
        }

        return Equals((TestClass)obj);
    }

    public override int GetHashCode()
    {
        return Test.GetHashCode();
    }

    protected bool Equals(TestClass other)
    {
        return Test == other.Test;
    }
}
```

### Expected behavior

The rule should not be triggered.

### Actual behavior

The rule reports that the class must be sealed or implement IEqualityComparer<T>.

### Known workarounds

None.

### Related information

Plugin csharp version 8.15.0.24505
SonarQube Developer EditionVersion 8.6 (build 39681)
dotnet 3.1
SonarScanner Plugin for Azure DevOps Version: 4.11.0

**Comment by pavel-mikula-sonarsource** [comment]
Hi @mpcleancode,

This case is a true positive for this rule as you're adding protected `Equals` method that is not override and is not part of `IEqualityComparer` interface. So any inheriting class can accidentally invoke it without being aware that it doesn't really do `Equals` for it's specific type, but rather it's base type.
